src/miral/reloading_config_file.cpp

```diff
@@ -0,0 +1,184 @@
+    inotify_fd{inotify_init1(IN_CLOEXEC)},
+    load_config{load_config},
+    filename{file.filename()},
+    directory{config_directory(file)},
+    directory_watch_descriptor{watch_descriptor(inotify_fd, directory)}
+{
+    register_handler(runner);
+
+    // With C++26 we should be able to use the optional directory as a range to
+    // initialize config_roots.  Until then, we'll just do it the long way...
+    std::vector<path> config_roots;
+
+    if (directory)
+    {
+        config_roots.push_back(directory.value());
+    }
+
+    if (auto config_dirs = getenv("XDG_CONFIG_DIRS"))
+    {
+        std::istringstream config_stream{config_dirs};
+        for (std::string config_root; getline(config_stream, config_root, ':');)
+        {
+            config_roots.push_back(config_root);
+        }
+    }
+    else
+    {
+        config_roots.push_back("/etc/xdg");
+    }
+
+    /* Read config file */
+    for (auto const& config_root : config_roots)
+    {
+        auto filepath = config_root / filename;
+        if (std::ifstream config_file{filepath})
+        {
+            load_config(config_file, filepath);
+            break;
+        }
+    }
+}
+
+miral::ReloadingConfigFile::ReloadingConfigFile(MirRunner& runner, path file, Loader load_config) :
+    self{std::make_shared<Self>(runner, file, load_config)}
+{
+}
+
+miral::ReloadingConfigFile::~ReloadingConfigFile() = default;
+
+void miral::ReloadingConfigFile::Self::register_handler(MirRunner& runner)
+{
+    if (directory_watch_descriptor)
+    {
+        fd_handle = runner.register_fd_handler(inotify_fd, [this] (int)
+        {
+            static size_t const sizeof_inotify_event = sizeof(inotify_event);
+
+            // A union ensures buffer is aligned for inotify_event
+            union
+            {
+                char buffer[sizeof_inotify_event + NAME_MAX + 1];
+                inotify_event unused [[maybe_unused]];
+            } inotify_magic;
+
+            auto const readsize = read(inotify_fd, &inotify_magic, sizeof(inotify_magic));
+            if (readsize < static_cast<ssize_t>(sizeof_inotify_event))
+            {
+                return;
+            }
+
+            auto raw_buffer = inotify_magic.buffer;
+            while (raw_buffer != inotify_magic.buffer + readsize)
+            {
+                // This is safe because inotify_magic.buffer is aligned and event.len includes padding for alignment
+                auto& event = reinterpret_cast<inotify_event&>(*raw_buffer);
+                if (event.mask & (IN_CLOSE_WRITE | IN_MOVED_TO) && event.wd == directory_watch_descriptor.value())
+                try
+                {
+                    if (event.name == filename)
+                    {
+                        auto const& file = directory.value() / filename;
+
+                        if (std::ifstream config_file{file})
+                        {
+                            load_config(config_file, file);
+                        }
```

Comment:
Else?

Reply:
Else we can't open the file, so no point trying to load a config from it

Reply:
Let's log it, then?